src/main/java/com/android/volley/RequestQueue.java

```diff
@@ -38,11 +41,56 @@
     // TODO: This should not be a generic class, because the request type can't be determined at
     // compile time, so all calls to onRequestFinished are unsafe. However, changing this would be
     // an API-breaking change. See also: https://github.com/google/volley/pull/109
+    @Deprecated // Use RequestEventListener instead.
     public interface RequestFinishedListener<T> {
         /** Called when a request has finished processing. */
         void onRequestFinished(Request<T> request);
     }
 
+    /** Request event types listeners will be notified about. */
+    @Retention(RetentionPolicy.SOURCE)
+    @IntDef({
+        RequestEvent.REQUEST_QUEUED,
+        RequestEvent.REQUEST_CACHE_LOOKUP_STARTED,
+        RequestEvent.REQUEST_CACHE_LOOKUP_FINISHED,
+        RequestEvent.REQUEST_PROCESSING_STARTED,
+        RequestEvent.REQUEST_PROCESSING_FINISHED,
+        RequestEvent.REQUEST_FINISHED
+    })
+    public @interface RequestEvent {
+        /** The request was added to the queue. */
+        public static final int REQUEST_QUEUED = 0;
+        /** The cache lookup started for the request. */
+        public static final int REQUEST_CACHE_LOOKUP_STARTED = 1;
+        /**
+         * The cache lookup finished for the request and cached response is delivered or request is
+         * queued for processing.
+         */
+        public static final int REQUEST_CACHE_LOOKUP_FINISHED = 2;
+        /** The processing (actual network access) started for the request. */
```

Comment:
as the fact that the parenthetical is necessary kind of gets at, "PROCESSING" seems unnecessarily vague here. What's happening is that the request is being sent over the network. Perhaps REQUEST_NETWORK_DISPATCH_STARTED/FINISHED would be clearer, and lead to a more natural Javadoc like "Network dispatch started for the request".